org.opent2t.sample.lamp.superpopular/com.hue.bulb/js/colour.js

```diff
@@ -0,0 +1,359 @@
+"use strict";
+
+const ChangeTolerance = 0.0001;
+const MaxHue = 360.0;
+const MaxColour = 255;
+
+/*
+ * Method to see if the given XY value is within the reach of the lamps.
+ *
+ * @param xyColour is a two-element object representing a point containing the X,Y value
+ * @param colourGamut   the color gamut of the light device.
+ * @return true if within reach, false otherwise.
+ */
+function isPointInLampsReach(xyColour, colourGamut){
+
+    var v1 = [colourGamut[1].x - colourGamut[0].x, colourGamut[1].y - colourGamut[0].y];
+    var v2 = [colourGamut[2].x - colourGamut[0].x, colourGamut[2].y - colourGamut[0].y];
+    var q = [xyColour.x - colourGamut[0].x, xyColour.y - colourGamut[0].y];
+
+    var s = product(q, v2) / product(v1, v2);
+    var t = product(v1, q) / product(v1, v2);
+
+    return ((s >= 0) && (t >= 0) && (s + t <= 1)) ? true : false;
+}
+
+/*
+ * Find the closest point on a line.
+ * This point will be within reach of the lamp.
+ *
+ * @param ptA the point where the line starts
+ * @param ptB the point where the line ends
+ * @param ptP the point which is close to a line.
+ * @return the point which is on the line.
+ */
+function getClosestPointToPoints(ptA, ptB, ptP){
+
+    var AP = {x: (ptP.x - ptA.x), y: (ptP.y - ptA.y)};
+    var AB = {x: (ptB.x - ptA.x), y: (ptB.y - ptA.y)};
+    var ab2 = (AB.x * AB.x) + (AB.y * AB.y);
+    var ap_ab = (AP.x * AB.x) + (AP.y * AB.y);
+
+    var t = ap_ab / ab2;
+
+    if (t < 0.0)
+    {
+        t = 0.0;
+    }
+    else if (t > 1.0)
+    {
+        t = 1.0;
+    }
+
+    return {x: (ptA.x + (AB.x * t)), y: (ptA.y + (AB.y * t))};
+}
+
+/*
+ * Find the distance between two points.
+ *
+ * @param pt1
+ * @param pt2
+ * @return the distance between point one and two
+ */
+function getDistanceBetweenTwoPoints(pt1, pt2){
+    var dx = pt1.x - pt2.x; // horizontal difference
+    var dy = pt1.y - pt2.y; // vertical difference
+    return Math.sqrt((dx * dx) + (dy * dy));
+}
+
+function product(pt1, pt2){
+    return ((pt1.x * pt2.y) - (pt1.y * pt2.x));
+}
+
+function getColourGamutForModel(modelId)
```

Comment:
>getColourGamutForModel [](start = 9, length = 22)

Can you put in a reference link to where these concepts like GamutA,B,C and supported colors etc are defined - eg [https://www.developers.meethue.com/documentation/core-concepts](https://www.developers.meethue.com/documentation/core-concepts)